enhancements/authentication/login-logout-events.md

```diff
@@ -0,0 +1,159 @@
+2. Extend oauth-server to capture login and login failure events.
+3. Extend oauth-apiserver to add user information when deleting `oauthaccesstoken` objects.
+4. Extend must-gather/gather_audit_logs to fetch oauth-server audit logs.
+
+### Non-Goals
+
+1. Hook into the existing API server [policy configuration](https://github.com/openshift/enhancements/blob/master/enhancements/kube-apiserver/audit-policy.md).
+2. Provide a fully featured audit subsystem in oauth-server compliant with API server policy framework.
+
+## Proposal
+
+We propose to add an `audit` struct to `oauth.config.openshift.io/v1` with a `profile` field which configures audit logging granularity.
+
+```go
+// OAuthSpec contains desired cluster auth configuration
+type OAuthSpec struct {
+	// identityProviders is an ordered list of ways for a user to identify themselves.
+	// When this list is empty, no identities are provisioned for users.
+	// +optional
+	IdentityProviders []IdentityProvider `json:"identityProviders,omitempty"`
+
+	// tokenConfig contains options for authorization and access tokens
+	TokenConfig TokenConfig `json:"tokenConfig"`
+
+	// templates allow you to customize pages like the login page.
+	// +optional
+	Templates OAuthTemplates `json:"templates"`
+
+	Audit OAuthAudit `json:"audit"`
+}
+
+type OAuthAudit struct {
+	Profile OAuthAuditProfileType `json:"profile,omitempty"`
+}
+
+type OAuthAuditProfileType string
+
+const (
+	// "None" disables audit logs.
+	OAuthNoneAuditProfileType AuditProfileType = "None"
+
+	// "WriteLoginEvents" logs login and login failure events.
+  // This is the default.
+	OAuthWriteLoginEventsProfileType AuditProfileType = "WriteLoginEvents"
+)
+```
+
+### User Stories
+
+### As a user I want to inspect login and login failure attempts
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect oauth-server audit logs.
+
+### As a user I want to inspect logout events
+
+Use `oc adm must-gather -- /usr/bin/gather_audit_logs` and inspect apiserver audit logs for deleted `oauthaccesstoken` resources.
+
+### Implementation Details/Notes/Constraints [optional]
+
+What are the caveats to the implementation? What are some important details that
+didn't come across above. Go in to as much detail as necessary here. This might
+be a good place to talk about core concepts and how they relate.
+
+## Design Details
+
+### Login and Login failure events
+
+To create login and login failure audit events differ slightly depending on the underlying identity provider. There are the following cases to consider:
+
+1. **Password-based identity providers**
+
+All password-based identity providers are using a [central location](https://github.com/openshift/oauth-server/blob/690499e76a0b242adb5ccc73f23cccc8a50b8788/pkg/server/login/login.go#L173) where login events are handled. Both login success and failure events will be emitted here.
```

Comment:
How do you plan to avoid a DOS attack where your disk resources would be depleted by an attacker forcing failure attempts audit in a loop? The oauth-server currently does not currently implement login attempt limits and adding it might be hard given we don't see the IP of the user but rather the one of the ingress.

Reply:
I agree, and I believe we briefly discussed this in the arch call as well. AFAIK the current audit event logging subsystem does not have rate limiting in general, I will add this as future work and as a non-goal.

Reply:
as discussed, we have that already "for free" as part of the apiserver default handler chain: https://github.com/openshift/oauth-server/blob/961295a3151da3fff9daa5ea0ab8c0bf92a7d7e6/vendor/k8s.io/apiserver/pkg/server/config.go#L807

I will add this as a paragraph in the OEP.